browser/farbling/farbling_tab_helper.cc

```diff
@@ -0,0 +1,169 @@
+  if (os_major_version == 10) {
+    switch (os_minor_version) {
+      case 9:
+      case 10:
+        return 5;
+      case 11:
+      case 12:
+      case 13:
+      case 14:
+      case 15:
+        return 6;
+    }
+  }
+  return 0;
+}
+#endif
+
+std::string GetUserAgentPlatform() {
+#if defined(OS_WIN)
+  return "";
+#elif defined(OS_MACOSX)
+  return "Macintosh; ";
+#elif defined(USE_X11)
+  return "X11; ";
+#elif defined(OS_ANDROID)
+  return "Linux; ";
+#elif defined(OS_POSIX)
+  return "Unknown; ";
+#endif
+}
+
+std::string GetMinimalProduct() {
+  return version_info::GetProductNameAndVersionForUserAgent();
+}
+
+std::string GetMinimalOSVersion() {
+  std::string os_version;
+#if defined(OS_WIN) || defined(OS_MACOSX)
+  int32_t os_major_version = 0;
+  int32_t os_minor_version = 0;
+  int32_t os_bugfix_version = 0;
+  base::SysInfo::OperatingSystemVersionNumbers(
+      &os_major_version, &os_minor_version, &os_bugfix_version);
+#endif
+#if defined(OS_MACOSX)
+  os_bugfix_version =
+      std::max(os_bugfix_version,
+               GetMinimumBugfixVersion(os_major_version, os_minor_version));
+#endif
+
+#if defined(OS_ANDROID)
+  std::string android_version_str = base::SysInfo::OperatingSystemVersion();
+  std::string android_info_str =
+      GetAndroidOSInfo(content::IncludeAndroidBuildNumber::Exclude,
+                       content::IncludeAndroidModel::Exclude);
+#endif
+
+  base::StringAppendF(&os_version,
+#if defined(OS_WIN)
+                      "%d.%d", os_major_version, os_minor_version
+#elif defined(OS_MACOSX)
+                      "%d_%d_%d", os_major_version, os_minor_version,
+                      os_bugfix_version
+#elif defined(OS_ANDROID)
+                      "%s%s", android_version_str.c_str(),
+                      android_info_str.c_str()
+#else
+                      ""
+#endif
+  );  // NOLINT
+  return os_version;
+}
+
+}  // namespace
+
+namespace brave {
+
+FarblingTabHelper::~FarblingTabHelper() = default;
+
+FarblingTabHelper::FarblingTabHelper(content::WebContents* web_contents)
+    : content::WebContentsObserver(web_contents) {}
+
+void FarblingTabHelper::DidStartNavigation(
+    content::NavigationHandle* navigation_handle) {
+  UpdateUserAgent(navigation_handle);
+}
+
+void FarblingTabHelper::DidRedirectNavigation(
+    content::NavigationHandle* navigation_handle) {
+  UpdateUserAgent(navigation_handle);
+}
+
+void FarblingTabHelper::UpdateUserAgent(
+    content::NavigationHandle* navigation_handle) {
+  if (!navigation_handle)
+    return;
+  content::WebContents* web_contents = navigation_handle->GetWebContents();
+  if (!web_contents)
+    return;
+  std::string ua = "";
+  // If shields is off or farbling is off, do not override.
+  Profile* profile = static_cast<Profile*>(web_contents->GetBrowserContext());
+  auto* map = HostContentSettingsMapFactory::GetForProfile(profile);
+  if (GetBraveShieldsEnabled(map, navigation_handle->GetURL())) {
```

Comment:
nit: I would use `&&` instead of nested `if` statements, it's just easier to read imo